src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServerLimits.cs

```diff
@@ -169,7 +171,14 @@ public int MaxRequestHeaderCount
         public TimeSpan KeepAliveTimeout
         {
             get => _keepAliveTimeout;
-            set => _keepAliveTimeout = value;
+            set
+            {
+                if (value <= TimeSpan.Zero && value != Timeout.InfiniteTimeSpan)
```

Comment:
InfiniteTimeSpan needs to be special cased in the timeout logic. I'd add a functional test that verifies that connections aren't killed immediate when InfiniteTimeSpan is configured.